File: 2020/23/p2.rb
class Node
  attr_reader :value
  attr_accessor :node

  def initialize(value)
    @value = value
  end

  def point(node)
    @node = node
  end
end

def target_from(val, max)
  val - 1 == 0 ? max : val - 1
end

$timings = Hash.new(0)

started = Time.now
input = File.open("input").lines.first.chomp.chars.map(&:to_i)
nodes = []
first_node = nil
last_node = nil
max_val = nil
input.each do |val|
  max_val = val if max_val.nil? || val > max_val
  node = Node.new(val)
  nodes[val] = node
  first_node ||= node
  last_node&.point(node)
  last_node = node
end
((max_val + 1)..1000000).each do |val|
  node = Node.new(val)
  nodes[val] = node
  last_node.point(node)
  last_node = node
end
max_val = 1000000
last_node.point(first_node)
$timings[:setup] = Time.now - started

game_loop = Time.now
current_node = first_node
10000000.times do |i|
  started = Time.now
  target_value = target_from(current_node.value, max_val)
  $timings[:first_target] += Time.now - started

  started = Time.now
  moving_values = []
  moving_nodes = current_node.node
  end_node = nil
  new_next = moving_nodes
  3.times do |_|
    end_node = new_next
    moving_values << new_next.value
    new_next = new_next.node
  end
  $timings[:move_selection] += Time.now - started
  started = Time.now
  until !moving_values.include?(target_value)
    target_value = target_from(target_value, max_val)
  end
  $timings[:retargeting] += Time.now - started

  started = Time.now
  current_node.node = new_next
  started_2 = Time.now
  insert_point = nodes[target_value]
  ended_2 = Time.now
  end_node.node = insert_point.node
  insert_point.node = moving_nodes
  $timings[:inserting] += started - started_2 + (Time.now - ended_2)
  $timings[:node_lookup] += ended_2 - started_2

  current_node = current_node.node
end
$timings[:game_loop] = Time.now - game_loop

current_node = nodes[1]

puts $timings
puts "After one comes #{current_node.node.value} and #{current_node.node.node.value} == #{current_node.node.value * current_node.node.node.value}"

File: 2020/23/p2.c
#include<stdio.h>
#include<stdlib.h>

void print_list(int* list, int current_val, int length) {
	printf("%d", current_val);
	int ind = current_val;
	for (int i = 1; i < length; i++) {
		ind = list[ind];
		printf("%d", ind);
	}
	printf("\n");
}

int target_from(int current_value, int max) {
	int dec = current_value - 1;
	return dec ? dec : max;
}

void main() {
	int size = 1000000;
	int* nodes = malloc(sizeof(int) * (size + 1));
	nodes[1] = 9;
	nodes[2] = 5;
	nodes[3] = 7;
	nodes[4] = 3;
	nodes[5] = 8;
	nodes[6] = 4;
	nodes[7] = 1;
	nodes[8] = 10;
	nodes[9] = 2;
	for (int i = 10; i <= size; i++) {
		nodes[i] = i + 1;
	}
	nodes[size] = 6;

	int current_val = 6;
	for (int i = 0; i < 10000000; i++) {
		int moving_val = nodes[current_val];
		int moving_values[3];
		moving_values[0] = moving_val;
		int new_next = moving_val;
		for (int j = 1; j < 3; j++) {
			new_next = nodes[new_next];
			moving_values[j] = new_next;
		}

		int target_value = target_from(current_val, size);
		while (moving_values[0] == target_value || moving_values[1] == target_value || moving_values[2] == target_value) {
			target_value = target_from(target_value, size);
		}

		int old_val = current_val;
		current_val = nodes[old_val] = nodes[new_next];
		nodes[new_next] = nodes[target_value];
		nodes[target_value] = moving_val;
	}

	int after_one = nodes[1];
	int after_that = nodes[after_one];

	printf("After one comes %d and %d == %lld\n", after_one, after_that, ((long long int) after_one) * after_that);
}
